File: backend/app/schemas/worklists.py
"""Lab worklist API schemas."""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from app.schemas.enums import (
    ContainerTopColor,
    ContainerType,
    PaymentStatus,
    PriorityLevel,
    SampleStatus,
    TestStatus,
)
from app.schemas.order import TestResultsDict


class WorklistPagination(BaseModel):
    page: int
    pageSize: int
    total: int
    totalPages: int
    hasNext: bool
    hasPrev: bool


class CollectionWorklistItem(BaseModel):
    sampleId: int
    orderId: int
    patientId: int
    patientName: str
    sampleType: str
    status: SampleStatus
    priority: PriorityLevel
    paymentStatus: PaymentStatus
    orderDate: datetime
    testCodes: list[str]
    isRecollection: bool
    originalSampleId: int | None = None
    originalSampleCollectedAt: datetime | None = None
    recollectionReason: str | None = None
    recollectionAttempt: int = 1
    blockedReason: str | None = None
    waitingHours: float
    turnaroundHours: int
    actualContainerType: ContainerType | None = None
    actualContainerColor: ContainerTopColor | None = None
    collectedAt: datetime | None = None
    collectedBy: str | None = None
    collectedVolume: float | None = None


class EntryWorklistItem(BaseModel):
    orderTestId: int
    orderId: int
    patientId: int
    patientName: str
    testCode: str
    testName: str
    sampleId: int | None
    sampleType: str
    priority: PriorityLevel
    status: TestStatus
    collectedAt: datetime | None
    orderDate: datetime
    waitingHours: float
    turnaroundHours: int
    isRetest: bool


class ValidationWorklistItem(BaseModel):
    orderTestId: int
    orderId: int
    patientId: int
    patientName: str
    testCode: str
    testName: str
    sampleType: str
    priority: PriorityLevel
    status: TestStatus
    sampleId: int | None = None
    sampleStatus: SampleStatus | None = None
    results: TestResultsDict | None = None
    flags: list[str] | None = None
    enteredBy: str | None = None
    referringPhysician: str | None = None
    isRetest: bool = False
    retestOfTestId: int | None = None
    retestNumber: int = 0
    resultEnteredAt: datetime | None
    orderDate: datetime
    waitingHours: float
    turnaroundHours: int
    hasCriticalValues: bool


class WorklistResponse(BaseModel):
    items: list
    pagination: WorklistPagination


class QueueAgeStats(BaseModel):
    oldestHours: float | None
    averageHours: float | None
    warningCount: int
    criticalCount: int


class BlockerSummary(BaseModel):
    paymentUnpaid: int
    retestPending: int
    recollectionWaiting: int
    total: int


class LabBoardCounts(BaseModel):
    collection: int
    entry: int
    validation: int
    supervisor: int


class AgeBuckets(BaseModel):
    fresh: int
    onTrack: int
    warning: int
    critical: int


class PriorityMix(BaseModel):
    urgent: int
    high: int
    medium: int
    low: int


class BoardAttentionItem(BaseModel):
    id: str
    stage: Literal["collection", "entry", "validation"]
    stageLabel: str
    orderId: int
    patientName: str
    priority: PriorityLevel
    waitingHours: float
    blockedReason: str | None = None
    blockedLabel: str | None = None
    queueTab: Literal["collection", "entry", "validation"]
    since: str
    workItemCount: int
    orderTestIds: list[int]
    attentionType: str


class LabBoardResponse(BaseModel):
    counts: LabBoardCounts
    queueAge: dict[str, QueueAgeStats]
    blockers: BlockerSummary
    health: Literal["healthy", "attention", "critical"]
    healthMessage: str
    suggestedTab: str | None
    ageBuckets: AgeBuckets
    priorityMix: PriorityMix
    attentionItems: list[BoardAttentionItem]
    attentionTotal: int
    totalActive: int
    computedAt: str | None = None

File: frontend/src/features/lab/api/worklists.ts
/**
 * Lab worklist API service — server-paginated queues.
 */
import { apiClient } from '@/lib/api/client';
import type {
  ContainerTopColor,
  ContainerType,
  PaymentStatus,
  PriorityLevel,
  SampleStatus,
  TestStatus,
} from '@/types';
import { useQuery } from '@tanstack/react-query';
import { useAuthStore } from '@/app/authStore';
import { queryKeys, cacheConfig } from '@/lib/query';
import { LAB_CONFIG } from '@/features/lab/constants';

export interface WorklistPagination {
  page: number;
  pageSize: number;
  total: number;
  totalPages: number;
  hasNext: boolean;
  hasPrev: boolean;
}

export interface CollectionWorklistItem {
  sampleId: number;
  orderId: number;
  patientId: number;
  patientName: string;
  sampleType: string;
  status: SampleStatus;
  priority: PriorityLevel;
  paymentStatus: PaymentStatus;
  orderDate: string;
  testCodes: string[];
  isRecollection: boolean;
  originalSampleId?: number | null;
  originalSampleCollectedAt?: string | null;
  recollectionReason?: string | null;
  recollectionAttempt?: number;
  blockedReason?: string | null;
  waitingHours: number;
  turnaroundHours: number;
  actualContainerType?: ContainerType | null;
  actualContainerColor?: ContainerTopColor | null;
  collectedAt?: string | null;
  collectedBy?: string | null;
  collectedVolume?: number | null;
}

export interface EntryWorklistItem {
  orderTestId: number;
  orderId: number;
  patientId: number;
  patientName: string;
  testCode: string;
  testName: string;
  sampleId?: number | null;
  sampleType: string;
  priority: PriorityLevel;
  status: TestStatus;
  collectedAt?: string | null;
  orderDate: string;
  waitingHours: number;
  turnaroundHours: number;
  isRetest: boolean;
}

export interface ValidationWorklistItem {
  orderTestId: number;
  orderId: number;
  patientId: number;
  patientName: string;
  testCode: string;
  testName: string;
  sampleType: string;
  priority: PriorityLevel;
  status: TestStatus;
  sampleId?: number | null;
  results?: Record<string, unknown> | null;
  flags?: string[] | null;
  enteredBy?: string | null;
  referringPhysician?: string | null;
  isRetest?: boolean;
  retestOfTestId?: number | null;
  retestNumber?: number;
  resultEnteredAt?: string | null;
  orderDate: string;
  waitingHours: number;
  turnaroundHours: number;
  hasCriticalValues: boolean;
}

interface WorklistParams {
  page?: number;
  pageSize?: number;
  search?: string;
}

function buildParams(params?: WorklistParams): Record<string, string> {
  const q: Record<string, string> = {};
  if (params?.page != null) q.page = String(params.page);
  if (params?.pageSize != null) q.pageSize = String(params.pageSize);
  if (params?.search) q.search = params.search;
  return q;
}

export const worklistsAPI = {
  getCollection(params?: WorklistParams) {
    return apiClient.get<{ items: CollectionWorklistItem[]; pagination: WorklistPagination }>(
      '/lab/worklists/collection',
      buildParams(params)
    );
  },

  getEntry(params?: WorklistParams) {
    return apiClient.get<{ items: EntryWorklistItem[]; pagination: WorklistPagination }>(
      '/lab/worklists/entry',
      buildParams(params)
    );
  },

  getValidation(params?: WorklistParams) {
    return apiClient.get<{ items: ValidationWorklistItem[]; pagination: WorklistPagination }>(
      '/lab/worklists/validation',
      buildParams(params)
    );
  },
};

/**
 * Lab worklist React Query hooks.
 */

interface WorklistHookParams {
  page?: number;
  pageSize?: number;
  search?: string;
}

export function useCollectionWorklist(params?: WorklistHookParams) {
  const { isAuthenticated, isLoading: isRestoring } = useAuthStore();
  const query = useQuery({
    queryKey: queryKeys.worklists.collection(params),
    queryFn: () => worklistsAPI.getCollection({ pageSize: 200, ...params }),
    enabled: isAuthenticated && !isRestoring,
    ...cacheConfig.dynamic,
    refetchInterval: LAB_CONFIG.TAB_COUNT_REFRESH_MS,
  });
  return {
    items: query.data?.items ?? [],
    pagination: query.data?.pagination,
    isLoading: query.isLoading,
    refetch: query.refetch,
  };
}

export function useEntryWorklist(params?: WorklistHookParams) {
  const { isAuthenticated, isLoading: isRestoring } = useAuthStore();
  const query = useQuery({
    queryKey: queryKeys.worklists.entry(params),
    queryFn: () => worklistsAPI.getEntry({ pageSize: 200, ...params }),
    enabled: isAuthenticated && !isRestoring,
    ...cacheConfig.dynamic,
    refetchInterval: LAB_CONFIG.TAB_COUNT_REFRESH_MS,
  });
  return {
    items: query.data?.items ?? [],
    pagination: query.data?.pagination,
    isLoading: query.isLoading,
    refetch: query.refetch,
  };
}

export function useValidationWorklist(params?: WorklistHookParams) {
  const { isAuthenticated, isLoading: isRestoring } = useAuthStore();
  const query = useQuery({
    queryKey: queryKeys.worklists.validation(params),
    queryFn: () => worklistsAPI.getValidation({ pageSize: 200, ...params }),
    enabled: isAuthenticated && !isRestoring,
    ...cacheConfig.dynamic,
    refetchInterval: LAB_CONFIG.TAB_COUNT_REFRESH_MS,
  });
  return {
    items: query.data?.items ?? [],
    pagination: query.data?.pagination,
    isLoading: query.isLoading,
    isError: query.isError,
    error: query.error,
    refetch: query.refetch,
  };
}
